Dockerfile

```diff
@@ -1,34 +1,36 @@
-FROM ruby:2.6.6
-RUN apt-get update -qq && apt-get upgrade -y
-RUN apt-get install -y build-essential nodejs && apt-get clean
-RUN gem install foreman
+# TODO: make this default to govuk-ruby once it's being pushed somewhere public
+# (unless we decide to use Bitnami instead)
+ARG base_image=ruby:2.6.6
 
+FROM $base_image AS builder
 # This image is only intended to be able to run this app in a production RAILS_ENV
-ENV RAILS_ENV production
-
-ENV GOVUK_APP_NAME whitehall
-ENV GOVUK_CONTENT_SCHEMAS_PATH /govuk-content-schemas
-ENV PORT 3020
-ENV DATABASE_URL mysql2://root:root@mysql/whitehall_development
-
-ENV APP_HOME /app
-RUN mkdir $APP_HOME
-
-WORKDIR $APP_HOME
-ADD Gemfile* $APP_HOME/
-ADD .ruby-version $APP_HOME/
-RUN bundle config set deployment 'true'
-RUN bundle config set without 'development test'
-RUN bundle install --jobs 4
-
-ADD . $APP_HOME
-
+ENV RAILS_ENV=production
+# TODO: have a separate build image which already contains the build-only deps.
+RUN apt-get update -qy && \
+    apt-get upgrade -y && \
+    apt-get install -y build-essential nodejs
+RUN mkdir /app
+WORKDIR /app
+COPY Gemfile Gemfile.lock .ruby-version /app/
+RUN bundle config set deployment 'true' && \
+    bundle config set without 'development test' && \
+    bundle install --jobs 4 --retry=2
+COPY . /app
+# TODO: We probably don't want assets in the image; remove this once we have a proper deployment process which uploads to (e.g.) S3.
 RUN GOVUK_ASSET_ROOT=https://assets.publishing.service.gov.uk \
   GOVUK_APP_DOMAIN=www.gov.uk \
   GOVUK_WEBSITE_ROOT=www.gov.uk \
   GOVUK_APP_DOMAIN_EXTERNAL=www.gov.uk \
   bundle exec rake shared_mustache:compile assets:precompile
 
-HEALTHCHECK CMD curl --silent --fail localhost:$PORT/healthcheck/ready || exit 1
-
-CMD foreman run web
+FROM $base_image
+ENV RAILS_ENV=production GOVUK_APP_NAME=whitehall
+# TODO: include nodejs in the base image (govuk-ruby).
+# TODO: apt-get upgrade in the base image
+RUN apt-get update -qy && \
+    apt-get upgrade -y && \
+    apt-get install -y nodejs
```

Comment:
Do you know if we definitely need nodejs at runtime? I anticipated it might only be needed to generate the assets.

Reply:
This is a good question. The short answer to that is i dont know for sure. There is yarn.lock and it was included in the original Dockerfile - which from what i can see as you mention is used to generate the assets. If i can be sure somehow that its not needed happy to omit it from this stage of the build.